contracts/hooks/examples/FullRange.sol

```diff
@@ -298,67 +252,63 @@ contract FullRange is BaseHook, ILockCallback {
-
-            if (position.tokensOwed1 > 0 || position.tokensOwed0 > 0) {
-                BalanceDelta balanceDelta = hookModifyPosition(
-                    key,
-                    IPoolManager.ModifyPositionParams({
-                        tickLower: MIN_TICK,
-                        tickUpper: MAX_TICK,
-                        liquidityDelta: -int256(int128(position.liquidity))
-                    })
-                );
-
-                uint160 newSqrtPriceX96 = uint160(
-                    FixedPointMathLib.sqrt(
-                        FullMath.mulDiv(
-                            uint128(-balanceDelta.amount1()), FixedPoint96.Q96, uint128(-balanceDelta.amount0())
-                        )
-                    ) * FixedPointMathLib.sqrt(FixedPoint96.Q96)
-                );
-
-                (uint160 sqrtPriceX96,,,,,) = poolManager.getSlot0(key.toId());
-
-                poolManager.swap(
-                    key,
-                    IPoolManager.SwapParams({
-                        zeroForOne: newSqrtPriceX96 < sqrtPriceX96,
-                        amountSpecified: MAX_INT,
-                        sqrtPriceLimitX96: newSqrtPriceX96
-                    })
-                );
-
-                uint128 liquidity = LiquidityAmounts.getLiquidityForAmounts(
-                    newSqrtPriceX96,
-                    TickMath.getSqrtRatioAtTick(MIN_TICK),
-                    TickMath.getSqrtRatioAtTick(MAX_TICK),
-                    uint256(uint128(-balanceDelta.amount0())),
-                    uint256(uint128(-balanceDelta.amount1()))
-                );
-
-                // reinvest everything
-                hookModifyPosition(
-                    key,
-                    IPoolManager.ModifyPositionParams({
-                        tickLower: MIN_TICK,
-                        tickUpper: MAX_TICK,
-                        liquidityDelta: int256(int128(liquidity))
-                    })
-                );
-
-                // update position
-                Position.Info memory posInfo = poolManager.getPosition(key.toId(), address(this), MIN_TICK, MAX_TICK);
-
-                position.feeGrowthInside0LastX128 = posInfo.feeGrowthInside0LastX128;
-                position.feeGrowthInside1LastX128 = posInfo.feeGrowthInside1LastX128;
-                position.tokensOwed0 = 0;
-                position.tokensOwed1 = 0;
-            }
+    function _rebalance(PoolKey calldata key, int256 paramsLiquidity) internal {
+        PoolId poolId = key.toId();
+        PoolInfo storage position = poolInfo[poolId];
+        if (position.owed && paramsLiquidity < 0) {
+            position.owed = false;
+
+            BalanceDelta balanceDelta = poolManager.modifyPosition(
+                key,
+                IPoolManager.ModifyPositionParams({
+                    tickLower: MIN_TICK,
+                    tickUpper: MAX_TICK,
+                    liquidityDelta: -int256(int128(position.liquidity))
+                })
```

Comment:
confused here....should we be taking out the liquidity as stated by `poolManager.getLiquidity()` which gets out of sync with our internal position.liquidity here as fees get accumulated? If what I said was true...we woudln't be swapping against a 0 liquidity pool....

Reply:
although it adds 300 gas to rebalancing, i assume syncing position.liquidity will add ~200 gas so maybe not a big deal